src/app/EventManagement.cpp

```diff
@@ -0,0 +1,867 @@
+    {
+        GetPriorityBuffer(aEventOptions.mpEventSchema->mPriority)->UpdateFirstLastEventTime(opts.mTimestamp);
+    }
+
+    opts.mUrgent       = aEventOptions.mUrgent;
+    opts.mpEventSchema = aEventOptions.mpEventSchema;
+
+    ctxt.mCurrentEventNumber = GetPriorityBuffer(opts.mpEventSchema->mPriority)->mLastEventNumber;
+    ctxt.mCurrentSystemTime  = GetPriorityBuffer(opts.mpEventSchema->mPriority)->mLastEventSystemTimestamp;
+
+    err = CalculateEventSize(apDelegate, &opts, requestSize);
+    SuccessOrExit(err);
+
+    // Ensure we have space in the in-memory logging queues
+    err = EnsureSpaceInCircularBuffer(requestSize);
+    SuccessOrExit(err);
+
+    err = ConstructEvent(&ctxt, apDelegate, &opts);
+    SuccessOrExit(err);
+
+    // Check the number of bytes written.  If the event is too large
+    // to be evicted from subsequent buffers, drop it now.
+    buffer = mpEventBuffer;
+    do
+    {
+        VerifyOrExit(buffer->GetQueueSize() >= writer.GetLengthWritten(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+        if (buffer->IsFinalDestinationForPriority(opts.mpEventSchema->mPriority))
+        {
+            break;
+        }
+        else
+        {
+            buffer = buffer->GetNextCircularEventBuffer();
+        }
+    } while (true);
+
+    mBytesWritten += writer.GetLengthWritten();
+
+exit:
+    ChipLogFunctError(err);
+    if (err != CHIP_NO_ERROR)
+    {
+        *mpEventBuffer = checkpoint;
+    }
+    else if (opts.mpEventSchema->mPriority >= CHIP_CONFIG_EVENT_GLOBAL_PRIORITY)
+    {
+        CircularEventBuffer * currentBuffer = GetPriorityBuffer(opts.mpEventSchema->mPriority);
+        aEventNumber                        = currentBuffer->VendEventNumber();
+        currentBuffer->UpdateFirstLastEventTime(opts.mTimestamp);
+
+#if CHIP_CONFIG_EVENT_LOGGING_VERBOSE_DEBUG_LOGS
+        ChipLogDetail(EventLogging,
+                      "LogEvent event number: %u schema priority: %u cluster id: 0x%x event id: 0x%x sys timestamp: 0x%" PRIx64,
+                      aEventNumber, opts.mpEventSchema->mPriority, opts.mpEventSchema->mClusterId, opts.mpEventSchema->mEventId,
+                      opts.mTimestamp.mValue);
+#endif // CHIP_CONFIG_EVENT_LOGGING_VERBOSE_DEBUG_LOGS
+
+        ScheduleFlushIfNeeded(opts.mUrgent);
+    }
+
+    return err;
+}
+
+CHIP_ERROR EventManagement::CopyEvent(const TLVReader & aReader, TLVWriter & aWriter, EventLoadOutContext * apContext)
+{
+    TLVReader reader;
+    TLVType containerType;
+    CopyAndAdjustDeltaTimeContext context(&aWriter, apContext);
+    const bool recurse = false;
+    CHIP_ERROR err     = CHIP_NO_ERROR;
+
+    reader.Init(aReader);
+    err = reader.EnterContainer(containerType);
+    SuccessOrExit(err);
+
+    err = aWriter.StartContainer(AnonymousTag, kTLVType_Structure, containerType);
+    SuccessOrExit(err);
+
+    err = TLV::Utilities::Iterate(reader, CopyAndAdjustDeltaTime, &context, recurse);
```

Comment:
may be easier to read as:

```cpp
 err = TLV::Utilities::Iterate(reader, CopyAndAdjustDeltaTime, &context, false /* recurse */);
```

so that I do not need to move my eyes around to see what values are set.

or for future PRs, replacing bools with enums would make code even faster to read:

```cpp
 err = TLV::Utilities::Iterate(reader, CopyAndAdjustDeltaTime, &context, TLV::kDoNotRecurse);
```

Reply:
same comment for below.